File: packages/containers/metadata-generator/src/dynamodb-writer.js
const { DynamoDBClient } = require('@aws-sdk/client-dynamodb');
const { DynamoDBDocumentClient, BatchWriteCommand, PutCommand } = require('@aws-sdk/lib-dynamodb');

class DynamoDBWriter {
  constructor(config) {
    this.tableName = config.tableName;
    this.client = DynamoDBDocumentClient.from(
      new DynamoDBClient({ region: config.region })
    );
    this.batchSize = 25; // DynamoDB batch write limit
    this.maxRetries = 3;
    this.baseRetryDelay = 1000; // 1 second
  }

  async batchWritePackages(packages) {
    console.log(`Starting batch write of ${packages.length} packages...`);
    
    let processedCount = 0;
    let errorCount = 0;
    
    // Process packages in batches
    for (let i = 0; i < packages.length; i += this.batchSize) {
      const batch = packages.slice(i, i + this.batchSize);
      
      try {
        await this.writeBatch(batch);
        processedCount += batch.length;
        
        if (processedCount % 100 === 0) {
          console.log(`Written ${processedCount}/${packages.length} packages...`);
        }
        
      } catch (error) {
        console.error(`Failed to write batch ${i}-${i + batch.length}:`, error.message);
        errorCount += batch.length;
        
        // Try to write individual items from failed batch
        for (const pkg of batch) {
          try {
            await this.writeIndividualPackage(pkg);
            processedCount++;
            errorCount--;
          } catch (individualError) {
            console.error(`Failed to write individual package ${pkg.packageName}:`, individualError.message);
          }
        }
      }
      
      // Add small delay to avoid throttling
      if (i + this.batchSize < packages.length) {
        await this.delay(100);
      }
    }
    
    console.log(`Batch write completed. Success: ${processedCount}, Errors: ${errorCount}`);
    
    if (errorCount > 0) {
      console.warn(`Warning: ${errorCount} packages failed to write`);
    }
  }

  async writeBatch(packages) {
    const putRequests = packages.map(pkg => ({
      PutRequest: {
        Item: this.serializePackage(pkg)
      }
    }));

    const params = {
      RequestItems: {
        [this.tableName]: putRequests
      }
    };

    return await this.executeWithRetry(async () => {
      const result = await this.client.send(new BatchWriteCommand(params));
      
      // Handle unprocessed items
      if (result.UnprocessedItems && Object.keys(result.UnprocessedItems).length > 0) {
        console.log(`Retrying ${Object.keys(result.UnprocessedItems[this.tableName] || {}).length} unprocessed items...`);
        
        const retryParams = {
          RequestItems: result.UnprocessedItems
        };
        
        await this.delay(1000); // Wait before retry
        return await this.client.send(new BatchWriteCommand(retryParams));
      }
      
      return result;
    });
  }

  async writeIndividualPackage(pkg) {
    const params = {
      TableName: this.tableName,
      Item: this.serializePackage(pkg)
    };

    return await this.executeWithRetry(async () => {
      return await this.client.send(new PutCommand(params));
    });
  }

  serializePackage(pkg) {
    // Ensure all required attributes are present and properly formatted
    return {
      packageName: pkg.packageName,
      version: pkg.version,
      attributePath: pkg.attributePath || '',
      description: pkg.description || '',
      homepage: pkg.homepage || '',
      license: pkg.license || '',
      platforms: pkg.platforms || [],
      maintainers: pkg.maintainers || [],
      broken: Boolean(pkg.broken),
      unfree: Boolean(pkg.unfree),
      lastUpdated: pkg.lastUpdated,
      hasEmbedding: Boolean(pkg.hasEmbedding)
    };
  }

  async executeWithRetry(operation) {
    let lastError;
    
    for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
      try {
        return await operation();
      } catch (error) {
        lastError = error;
        
        // Check if error is retryable
        if (this.isRetryableError(error) && attempt < this.maxRetries) {
          const delay = this.baseRetryDelay * Math.pow(2, attempt - 1); // Exponential backoff
          console.log(`Attempt ${attempt} failed, retrying in ${delay}ms...`);
          await this.delay(delay);
          continue;
        }
        
        throw error;
      }
    }
    
    throw lastError;
  }

  isRetryableError(error) {
    const retryableErrors = [
      'ProvisionedThroughputExceededException',
      'ThrottlingException',
      'RequestLimitExceeded',
      'InternalServerError',
      'ServiceUnavailable'
    ];
    
    return retryableErrors.some(errorType => 
      error.name === errorType || 
      error.message.includes(errorType)
    );
  }

  delay(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

module.exports = { DynamoDBWriter };

File: packages/containers/metadata-generator/packages/containers/metadata-generator/src/dynamodb_writer.py
import logging
from typing import Any, Dict, List

import boto3
from botocore.exceptions import ClientError


logger = logging.getLogger(__name__)


class DynamoDBWriter:
    def __init__(self, table_name: str, region: str) -> None:
        self.table_name = table_name
        self.region = region
        self.dynamodb = boto3.resource("dynamodb", region_name=region)
        self.table = self.dynamodb.Table(table_name)
        self.batch_size = 25  # DynamoDB batch write limit

    def batch_write_packages(self, packages: List[Dict[str, Any]]) -> None:
        logger.info("Starting batch write of %d packages...", len(packages))

        processed = 0
        errors = 0

        for i in range(0, len(packages), self.batch_size):
            batch = packages[i : i + self.batch_size]
            try:
                self._write_batch(batch)
                processed += len(batch)
                if processed % 100 == 0:
                    logger.info("Written %d/%d packages...", processed, len(packages))
            except Exception as e:
                logger.error("Failed to write batch %d-%d: %s", i, i + len(batch), str(e))
                # attempt individual writes
                for pkg in batch:
                    try:
                        self._write_single(pkg)
                        processed += 1
                    except Exception as ie:
                        errors += 1
                        logger.warning(
                            "Failed to write individual package %s: %s",
                            pkg.get("packageName", "<unknown>"),
                            str(ie),
                        )

        logger.info("Batch write completed. Success: %d, Errors: %d", processed, errors)
        if errors > 0:
            logger.warning("%d packages failed to write", errors)

    def _write_batch(self, items: List[Dict[str, Any]]) -> None:
        # boto3 batch_writer handles retries for unprocessed items
        with self.table.batch_writer(overwrite_by_pkeys=["packageName", "version"]) as batch:
            for item in items:
                batch.put_item(Item=self._serialize(item))

    def _write_single(self, item: Dict[str, Any]) -> None:
        self.table.put_item(Item=self._serialize(item))

    def _serialize(self, pkg: Dict[str, Any]) -> Dict[str, Any]:
        # Ensure presence and proper types; DynamoDB via boto3 will map dicts/lists appropriately
        return {
            "packageName": pkg.get("packageName"),
            "version": pkg.get("version"),
            "attributePath": pkg.get("attributePath", ""),
            "description": pkg.get("description", ""),
            "longDescription": pkg.get("longDescription", ""),
            "homepage": pkg.get("homepage", ""),
            "license": pkg.get("license"),
            "platforms": pkg.get("platforms", []),
            "maintainers": pkg.get("maintainers", []),
            "broken": bool(pkg.get("broken", False)),
            "unfree": bool(pkg.get("unfree", False)),
            "available": bool(pkg.get("available", True)),
            "insecure": bool(pkg.get("insecure", False)),
            "unsupported": bool(pkg.get("unsupported", False)),
            "mainProgram": pkg.get("mainProgram", ""),
            "position": pkg.get("position", ""),
            "outputsToInstall": pkg.get("outputsToInstall", []),
            "lastUpdated": pkg.get("lastUpdated"),
            "hasEmbedding": bool(pkg.get("hasEmbedding", False)),
        }

